### PR TITLE
Run smart contract via blockchain hook

### PR DESCRIPTION
**Pull request by matei-p**

### CODE REVIEW

File: arwen/host/execution.go [hunk 1]
```diff
@@ -254,6 +254,23 @@ func (host *vmHost) ExecuteOnDestContext(input *vmcommon.ContractCallInput) (vmO
 	scExecutionInput := input
 
 	blockchain := host.Blockchain()
+
+	isSmartContract := blockchain.IsSmartContract(input.RecipientAddr)
```
Comment (sasurobert) [inline arwen/host/execution.go:258]: I think this code should be more inside exeucteOnDestContextNoBuiltInFunction.

File: arwen/host/execution.go [hunk 2]
```diff
@@ -429,6 +446,14 @@ func (host *vmHost) ExecuteOnSameContext(input *vmcommon.ContractCallInput) erro
 
 	managedTypes, blockchain, metering, output, runtime, _, _ := host.GetContexts()
 
+	vmType, err := vmcommon.ParseVMTypeFromContractAddress(input.RecipientAddr)
+	if err != nil {
+		return err
+	}
+	if !bytes.Equal(runtime.GetVMType(), vmType) {
+		return vmcommon.ErrInvalidVMType
```
Comment (sasurobert) [inline arwen/host/execution.go:454]: do we really need this ?

File: arwen/host/execution.go (hunk 1 above)
Comment (sasurobert) [reply arwen/host/execution.go:258]: Wouldn't it be better to have the code deep inside func (host *vmHost) execute(input *vmcommon.ContractCallInput) error ?

File: arwenmandos/exec.go [hunk 3]
```diff
@@ -101,6 +103,17 @@ func (ae *ArwenTestExecutor) GetVM() vmi.VMExecutionHandler {
 
 // GetVMHost returns de vm Context from the vm context map
 func (ae *ArwenTestExecutor) GetVMHost() arwen.VMHost {
+	skipStackLevels := 2
```
Comment (sasurobert) [inline arwenmandos/exec.go:106]: this is some hardcoded stuff. Can't we do better ?

File: mandos-go/expression/interpreter/interpreter.go [hunk 4]
```diff
@@ -181,6 +183,15 @@ func (ei *ExprInterpreter) InterpretString(strRaw string) ([]byte, error) {
 	return ei.interpretNumber(strRaw, 0)
 }
 
+// GetVMType yields the configured VM type, which is used for generating SC addresses.
+// Will yield default value [0, 0] is not explicitly configured.
+func (ei *ExprInterpreter) GetVMType() *[core.VMTypeLen]byte {
+	if ei.VMType == nil {
+		ei.VMType = &[core.VMTypeLen]byte{0, 0}
```
Comment (sasurobert) [inline mandos-go/expression/interpreter/interpreter.go:190]: I think this should yield WASM VM type by default.

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: arwen/host/execution.go (hunk 1 above)
Comment (matei-p) [reply arwen/host/execution.go]: refactored and used a logic similar with the one from builtin functions

File: arwen/host/execution.go (hunk 2 above)
Comment (matei-p) [reply arwen/host/execution.go]: removed it

File: arwenmandos/exec.go (hunk 3 above)
Comment (matei-p) [reply arwenmandos/exec.go:106]: @andrei-marinica

File: mandos-go/expression/interpreter/interpreter.go (hunk 4 above)
Comment (matei-p) [reply mandos-go/expression/interpreter/interpreter.go:190]: @andrei-marinica

Comment (andrei-marinica) [reply mandos-go/expression/interpreter/interpreter.go:190]: This component has no access to the VM. I made it editable, but there is no mechanism yet to set it.
There is a `var DefaultVMType = []byte{0xF, 0xF}`, should I rather have that as default?

File: arwenmandos/exec.go (hunk 3 above)
Comment (andrei-marinica) [reply arwenmandos/exec.go:106]: I don't really know what that variable means. This is the mandos executor, but I guess we could extract to a constant.

**Review by andrei-marinica — APPROVED** [review]

File: arwenmandos/exec.go (hunk 3 above)
Comment (camilbancioiu) [reply arwenmandos/exec.go:108]: Deleted, this is unnecessary. This code appears to have been copied from the old `GetVMHost()`, when it had VM hook logs. But here, `arwenmandos.GetVMHost()` was not called from within VM hooks, so the whole logging code here did nothing useful.

**Review by camilbancioiu — DISMISSED** [review]

**Review by camilbancioiu — APPROVED** [review]

**Review by andrei-marinica — APPROVED** [review]

**Pull request merged**